File: mobile_client/lib/services/home_page/home_lobby_service.dart
import 'package:collection/collection.dart';
import 'package:flutter/foundation.dart';
import 'package:mobile_client/constants/socket_events/home_event.dart';
import 'package:mobile_client/models/home_page/active_game.dart';
import 'package:mobile_client/models/home_page/new_player_count.dart';
import 'package:mobile_client/models/home_page/new_room_state.dart';
import 'package:mobile_client/services/websocket_service.dart';

class HomeLobbyService extends ChangeNotifier {
  List<ActiveGame> activeGames = [];

  final WebSocketService _webSocketService;
  HomeLobbyService(this._webSocketService);

  void joinHomeLobby() {
    activeGames = [];
    _setupUpdateGameListListener();
    _setupPlayerCountListener();
    _setupRemoveGameListener();
    _setupUpdateRoomStateListener();

    _webSocketService.send(HomeEvent.joinHomeLobby.name);
  }

  void leaveHomeLobby() {
    _webSocketService.removeAllListeners(HomeEvent.updateGameList.name);
    _webSocketService.removeAllListeners(HomeEvent.updatePlayerCount.name);
    _webSocketService.removeAllListeners(HomeEvent.updateRoomState.name);
    _webSocketService.removeAllListeners(HomeEvent.removeGame.name);
    _webSocketService.send(HomeEvent.leaveHomeLobby.name);
    activeGames = [];
    notifyListeners();
  }

  void _setupUpdateGameListListener() {
    _webSocketService.on<List<dynamic>>(HomeEvent.updateGameList.name,
        (newActiveGames) {
      activeGames = [
        ...activeGames,
        ...newActiveGames.map((a) => ActiveGame.fromJson(a))
      ];
      notifyListeners();
    });
  }

  void _setupPlayerCountListener() {
    _webSocketService.on<Map<String, dynamic>>(HomeEvent.updatePlayerCount.name,
        (json) {
      final playerCountChange = NewPlayerCount.fromJson(json);
      var changedGame = activeGames.firstWhereOrNull(
          (activeGame) => activeGame.roomId == playerCountChange.roomId);
      if (changedGame == null) return;

      changedGame.playerCount += playerCountChange.playerCount;
      notifyListeners();
    });
  }

  void _setupRemoveGameListener() {
    _webSocketService.on<String>(HomeEvent.removeGame.name, (roomId) {
      activeGames = activeGames
          .where((activeGame) => activeGame.roomId != roomId)
          .toList();
      notifyListeners();
    });
  }

  void _setupUpdateRoomStateListener() {
    _webSocketService.on<Map<String, dynamic>>(HomeEvent.updateRoomState.name,
        (json) {
      final newRoomState = NewRoomState.fromJson(json);

      var changedGame = activeGames.firstWhereOrNull(
          (activeGame) => activeGame.roomId == newRoomState.roomId);
      if (changedGame == null) return;

      changedGame.roomState = newRoomState.roomState;
      notifyListeners();
    });
  }
}

File: desktop-client/src/app/home/services/home-lobby/home-lobby.service.ts
import { Injectable } from '@angular/core';
import { WebsocketService } from '@app/shared/services/websocket/websocket.service';
import { ActiveGame, NewPlayerCount, NewRoomState } from '@common/interfaces/active-game';
import { HomeEvent } from '@common/socket-events/home-event';

@Injectable({
    providedIn: 'root',
})
export class HomeLobbyService {
    activeGames: ActiveGame[] = [];
    constructor(private readonly socketService: WebsocketService) {}

    joinHomeLobby() {
        this.activeGames = [];
        this.setupUpdateGameListListener();
        this.setupPlayerCountListener();
        this.setupRemoveGameListener();
        this.setupUpdateRoomStateListener();

        this.socketService.send(HomeEvent.JOIN_HOME_LOBBY);
    }

    leaveHomeLobby() {
        this.socketService.removeAllListeners(HomeEvent.UPDATE_GAME_LIST);
        this.socketService.removeAllListeners(HomeEvent.UPDATE_PLAYER_COUNT);
        this.socketService.removeAllListeners(HomeEvent.UPDATE_ROOM_STATE);
        this.socketService.removeAllListeners(HomeEvent.REMOVE_GAME);
        this.socketService.send(HomeEvent.LEAVE_HOME_LOBBY);
        this.activeGames = [];
    }

    private setupUpdateGameListListener() {
        this.socketService.on(HomeEvent.UPDATE_GAME_LIST, (activeGames?: ActiveGame[]) => {
            if (activeGames) this.activeGames = this.activeGames.concat(activeGames);
        });
    }

    private setupPlayerCountListener() {
        this.socketService.on(HomeEvent.UPDATE_PLAYER_COUNT, (playerCountChange?: NewPlayerCount) => {
            if (!playerCountChange) return;
            const changedGame = this.activeGames.find((activeGame) => activeGame.roomId === playerCountChange.roomId);
            if (!changedGame) return;

            changedGame.playerCount += playerCountChange.playerCount;
        });
    }

    private setupRemoveGameListener() {
        this.socketService.on(HomeEvent.REMOVE_GAME, (roomId?: string) => {
            if (!roomId) return;
            roomId = roomId.toString();

            this.activeGames = this.activeGames.filter((activeGame) => activeGame.roomId !== roomId);
        });
    }

    private setupUpdateRoomStateListener() {
        this.socketService.on(HomeEvent.UPDATE_ROOM_STATE, (newRoomState?: NewRoomState) => {
            if (!newRoomState) return;
            const changedGame = this.activeGames.find((activeGame) => activeGame.roomId === newRoomState.roomId);
            if (!changedGame) return;

            changedGame.roomState = newRoomState.roomState;
        });
    }
}
